### PR TITLE
tools: add tcptracer

### PR DESCRIPTION
**Pull request by iaguis**
This allows tracking TCP connections by tracking TCP connects, closes
and accepts.

This is different from existing tools like tcpconnect or tcpaccept in
that:

* It includes more information like network namespace or source ports
for tcpconnects or remote ports for tcpaccepts
* It traces tcp_close allowing to see when a connection ends
* It only shows information about established connections

### CODE REVIEW

**Comment by brendangregg** [comment]
Thanks; some quick comments:

- This isn't an /example, it's a /tool. Please see CONTRIBUTING_SCRIPTS.md.
- It uses 5 kernel dynamic tracepoints just for IPv4. This makes it especially brittle. tcpconnect uses 2, and tcpaccept uses 1, and they both cover IPv4 & IPv6.

This is the kind of tool I'd rather wait until we have TCP tracepoints to attempt -- speaking from experience. My tcpsnoop script from Solaris 10 used several dynamic kernel tracepoints, and it broke at least 7 times, and I ended up with a lot of upset end users. Once we had static tracepoints, I was able to write such tools once and they never broke again. We need them on Linux. And, should they arrive, supporting IPv6 should be much easier -- this would become a "tcptracer" tool.

Although, Solaris is obviously not Linux. The real question here is how likely these are to change:

- int kprobe__tcp_v4_connect(struct pt_regs *ctx, struct sock *sk)
- int kprobe__tcp_close(struct pt_regs *ctx, struct sock *sk)
- int kretprobe__tcp_v4_connect(struct pt_regs *ctx)
- int kretprobe__tcp_close(struct pt_regs *ctx)
- int kretprobe__inet_csk_accept(struct pt_regs *ctx)

We can add details like ports and namespaces to the existing tcpconnect/tcpaccept. And there's another ticket to implement a tcplife tool, for tcp_close() tracing.

**Comment by brendangregg** [comment]
The output is also over 80 chars, by _default_, when it doesn't need to be. That breaks a lot of things, and should only be done if it's unavoidable.

If I were writing it, I'd have made the TYPE column a single character "T" with "C" == connect, etc. NETNS only printed with a -n or -v option. Truncated COMM a little (12 chars) if need be. Etc.

**Comment by 4ast** [comment]
> Although, Solaris is obviously not Linux. The real question here is how likely these are to change:

    int kprobe__tcp_v4_connect(struct pt_regs *ctx, struct sock *sk)
    int kprobe__tcp_close(struct pt_regs *ctx, struct sock *sk)
    int kretprobe__tcp_v4_connect(struct pt_regs *ctx)
    int kretprobe__tcp_close(struct pt_regs *ctx)
    int kretprobe__inet_csk_accept(struct pt_regs *ctx)

These funcs and their args are highly unlikely to change any time soon.
I think it's ok to go with kprobe approach for now, since there is no one actively proposing tracepoints for tcp stack and it's not clear what kind of push back there will be.

**Comment by brendangregg** [comment]
Thanks @4ast. Yes, I really want TCP tracepoints, and maybe I'll just have to do propose them, but that's all in the future at the moment.

@iaguis So I'd check CONTRIBUTING_SCRIPTS.md, which will mean pep8 styling, an examples file, and a man page. Try to adjust the output width to keep it <80, which probably means putting the NETNS column in a verbose mode (-v). After all, if you run it with -n and have selected a NETNS, then you don't need to print it out by default.

I'd also consider changing -n to -N. -n is often used to match a process name, and that's a reasonable addition to this tool at some point ("tcpv4tracer java").

**Comment by brendangregg** [comment]
Also, why do you bother with the kretprobe of tcp_close()? Is something not valid on the kprobe?

**Comment by iaguis** [comment]
Thanks everybody for the quick responses!

> Also, why do you bother with the kretprobe of tcp_close()? Is something not valid on the kprobe?

You're right, the kprobe should be enough. Thanks!

I'm sorry I haven't read CONTRIBUTING_SCRIPTS.md before submitting this, I'll rework this PR taking that and https://github.com/iovisor/bcc/pull/762#issuecomment-254598664 into account.

**Comment by goldshtn** [comment]
@iaguis This PR is still open -- is it ready for another review?

**Comment by iaguis** [comment]
> @iaguis This PR is still open -- is it ready for another review?

I'll try to rebase it and update by the end of the week.

**Comment by 4ast** [comment]
@iaguis waiting for rebase and addressing @brendangregg comments re: contributing_scripts.md

**Comment by goldshtn** [comment]
@iaguis Ping? :)

**Comment by iaguis** [comment]
Rebased, addressed comments (I think I didn't forget any).

**Comment by goldshtn** [comment]
[buildbot, ok to test]

**Comment by goldshtn** [comment]
[buildbot, test this please]

**Comment by goldshtn** [comment]
Copying @brendangregg @4ast to take a look.

**Comment by brendangregg** [comment]
It's pretty long, at 649 lines, which would make it the 3rd longest tool in bcc (coming after the multi-tools trace and argdist). My longest tool is tcplife, at 360 lines, and my average is 194. Can the header include something like?:

```
# You should generally try to avoid writing long scripts that measure multiple functions and
# walk multiple kernel structures, as they will be a burden to maintain as the kernel changes.
# The following code should be replaced, and simplified, when static TCP probes exist.
```

My hope is to discourage others copying this and making it longer/worse. We need static TCP tracepoints, which should cut down a lot of these lines, and make the script stable.

BTW:

```
# ./tcptracer.py -h
usage: tcptracer.py [-h] [-p PID] [-N NETNS] [-v]

Trace TCP connections

optional arguments:
  -h, --help            show this help message and exit
  -p PID, --pid PID     trace this PID only
  -N NETNS, --netns NETNS
                        trace this Network Namespace only
  -v, --verbose         include Network Namespace in the output
```

It doesn't do that much right now, either. When people start adding options, this tool will quickly become the biggest in bcc.

Other comments...

-v has a \n in the header:

```
# ./tcptracer.py -v
Tracing TCP established connections. Ctrl-C to end.
TYPE         PID    COMM             IP SADDR            DADDR            SPORT  DPORT 
NETNS   
```

And the T column:

```
./tcptracer.py 
Tracing TCP established connections. Ctrl-C to end.
T  PID    COMM             IP SADDR            DADDR            SPORT  DPORT 
AC 1171   sshd             4  100.66.11.247    100.72.22.55     22     27692 
CL 19801  sshd             4  100.66.11.247    100.72.22.55     22     27692 
```

When I see two chars, I think flags (like the RWBS field from the block tracepoints). I'd consider changing it to a single character (which matches the header, "T"):

- A: accept
- C: connect
- X: close

or, to be TCP-terminology pedantic:

- A: active open (what at the socket layer would be a connect())
- P: passive open (what at the socket layer would be an accept())
- C: close

Maybe the A/C/X is better, as those terms are more commonly used.

**Comment by iaguis** [comment]
Thanks for the comments. Updated.

**Comment by goldshtn** [comment]
[buildbot, ok to test]

File: tools/tcptracer_example.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# ./tcptracer
+T  PID    COMM             IP SADDR            DADDR            SPORT  DPORT
+C 2389    telnet           4  192.168.1.2      192.168.1.1      53404  23
+C 23323   curl             6  [::1]            [::1]            36226  80
+X 2389    telnet           4  192.168.1.2      192.168.1.1      53404  23
+A 23278   nc               6  [::1]            [::1]            80     36226
+X 23323   curl             6  [::1]            [::1]            36226  80
+X 23278   nc               6  [::1]            [::1]            80     36226
+A 15195   nc               4  10.202.210.1     10.202.109.12    8080   43904
+X 15195   nc               4  10.202.210.1     10.202.109.12    8080   43904
```
Comment (goldshtn) [inline tools/tcptracer_example.txt]: Is the PID column header really misaligned now with the first digit of the pid?

Comment (iaguis) [inline tools/tcptracer_example.txt]: Ooops. No, that was me being off-by-one when editing the file. Let me re-run the test scenario.

**Comment by goldshtn** [comment]
[buildbot, ok to test]

**Comment by drzaeus77** [comment]
@goldshtn 
I updated one more setting on the buildbot, your [] should work next time.

**Comment by brendangregg** [comment]
LGTM

**Comment by brendangregg** [comment]
[buildbot, ok to test]

**Pull request merged**